File: internal/rules/no_meaningless_void_operator/no_meaningless_void_operator_test.go
package no_meaningless_void_operator

import (
	"testing"

	"github.com/typescript-eslint/tsgolint/internal/rule_tester"
	"github.com/typescript-eslint/tsgolint/internal/rules/fixtures"
	"github.com/typescript-eslint/tsgolint/internal/utils"
)

func TestNoMeaninglessVoidOperatorRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &NoMeaninglessVoidOperatorRule, []rule_tester.ValidTestCase{
		{Code: `
(() => {})();

function foo() {}
foo(); // nothing to discard

function bar(x: number) {
  void x;
  return 2;
}
void bar(); // discarding a number
    `},
		{Code: `
function bar(x: never) {
  void x;
}
    `},
	}, []rule_tester.InvalidTestCase{
		{
			Code:   "void (() => {})();",
			Output: []string{" (() => {})();"},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "meaninglessVoidOperator",
					Line:      1,
					Column:    1,
				},
			},
		},
		{
			Code: `
function foo() {}
void foo();
      `,
			Output: []string{`
function foo() {}
 foo();
      `,
			},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "meaninglessVoidOperator",
					Line:      3,
					Column:    1,
				},
			},
		},
		{
			Code: `
function bar(x: never) {
  void x;
}
      `,
			Options: NoMeaninglessVoidOperatorOptions{CheckNever: utils.Ref(true)},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "meaninglessVoidOperator",
					Line:      3,
					Column:    3,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "removeVoid",
							Output: `
function bar(x: never) {
   x;
}
      `,
						},
					},
				},
			},
		},
	})
}

File: packages/rslint-test-tools/tests/typescript-eslint/rules/no-meaningless-void-operator.test.ts
import { describe, test, expect } from '@rstest/core';
import { RuleTester } from '@typescript-eslint/rule-tester';

import { getFixturesRootDir } from '../RuleTester';

const rootDir = getFixturesRootDir();
const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootDir,
    },
  },
});

describe('no-meaningless-void-operator', () => {
  test('rule tests', () => {
    ruleTester.run('no-meaningless-void-operator', {
      valid: [
        `
(() => {})();

function foo() {}
foo(); // nothing to discard

function bar(x: number) {
  void x;
  return 2;
}
void bar(); // discarding a number
    `,
        `
function bar(x: never) {
  void x;
}
    `,
      ],
      invalid: [
        {
          code: 'void (() => {})();',
          errors: [
            {
              column: 1,
              line: 1,
              messageId: 'meaninglessVoidOperator',
            },
          ],
          output: '(() => {})();',
        },
        {
          code: `
function foo() {}
void foo();
      `,
          errors: [
            {
              column: 1,
              line: 3,
              messageId: 'meaninglessVoidOperator',
            },
          ],
          output: `
function foo() {}
foo();
      `,
        },
        {
          code: `
function bar(x: never) {
  void x;
}
      `,
          errors: [
            {
              column: 3,
              line: 3,
              messageId: 'meaninglessVoidOperator',
              suggestions: [
                {
                  messageId: 'removeVoid',
                  output: `
function bar(x: never) {
  x;
}
      `,
                },
              ],
            },
          ],
          options: [{ checkNever: true }],
          output: null,
        },
      ],
    });
  });
});
